### ISSUE TITLE
Questions about whether the teacher network is trained or not

### ISSUE THREAD

**Issue by Mamduh-k** [lhoyer/DAFormer#59]
Dear author, is the ema_model trainable except for updating through ema?

**Comment by Mamduh-k** [comment]
Also, I would like to know why there is no gradient backpropagation to the teacher network.

**Comment by lhoyer** [comment]
The teacher is only upgraded by the exponential moving average update of the network weights to implement a temporal ensemble. Therefore, there are no gradients backpropagated into the teacher. Please, have a look at https://arxiv.org/pdf/1703.01780.pdf for further details.

**Comment by Mamduh-k** [comment]
Thank you for your answer, but I wonder if the teacher network does not seem to have gradient back into it during training, so why is "detach()" needed?

**Comment by lhoyer** [comment]
The `mix_loss.backward()` uses pseudo-labels predicted by the teacher network. Without the detach(), the gradients could be backpropagated into the teacher.
